File: docs/sphinx/snippets/python/operators/spin_op_creation.py
from cudaq import spin
import cudaq # For observe, kernel, x
import sys # For sys.stderr and sys.exit

# [Begin SpinOp Creation Python]
# The following lines define the spin_op `h` as shown in the documentation.
h_py_example = 5.907 - 2.1433 * spin.x(0) * spin.x(1) - 2.1433 * spin.y(
    0) * spin.y(1) + .21829 * spin.z(0) - 6.125 * spin.z(1)
# [End SpinOp Creation Python]

# [Begin SpinOp Creation Python Execution]
if __name__ == "__main__":
    print(f"Python Spin Operator h_py_example:\n{h_py_example}")

    if h_py_example.get_term_count() == 0:
        print("Error: Python Spin operator h_py_example has no terms.", file=sys.stderr)
        sys.exit(1) # Indicate failure

    # Example of using the spin_op in an observe call with a simple ansatz.
    # The spin_op uses qubits 0 and 1, so we need at least 2 qubits.
    @cudaq.kernel
    def ansatz_py():
        q = cudaq.qvector(2)
        x(q[0]) # A simple operation on one of the qubits.

    energy = cudaq.observe(ansatz_py, h_py_example)
    print(f"Observed energy (Python): {energy.expectation()}")
# [End SpinOp Creation Python Execution]

File: docs/sphinx/snippets/cpp/operators/spin_op_creation.cpp
#include <cudaq.h>
#include <cudaq/spin_op.h>
#include <iostream> // For std::cout and std::cerr

// [Begin SpinOp Creation C++]
// The following lines define the spin_op `h` as shown in the documentation.
auto h_cpp_example =
    5.907 - 2.1433 * cudaq::spin_op::x(0) * cudaq::spin_op::x(1) -
    2.1433 * cudaq::spin_op::y(0) * cudaq::spin_op::y(1) +
    .21829 * cudaq::spin_op::z(0) - 6.125 * cudaq::spin_op::z(1);
// [End SpinOp Creation C++]

int main() {
  // [Begin SpinOp Creation C++ Execution]
  // This part is for testing and making the snippet runnable.
  // It won't be included in the RST by the literalinclude directive.
  std::cout << "C++ Spin Operator h_cpp_example:\n"
            << h_cpp_example.to_string() << std::endl;

  if (h_cpp_example.get_term_count() == 0) {
    std::cerr << "Error: C++ Spin operator h_cpp_example has no terms."
              << std::endl;
    return 1; // Indicate failure
  }

  // Example of using the spin_op in an observe call with a simple ansatz.
  // The spin_op uses qubits 0 and 1, so we need at least 2 qubits.
  struct ansatz_cpp {
    void operator()() __qpu__ {
      cudaq::qarray<2> q;
      x(q[0]); // A simple operation on one of the qubits.
    }
  };

  double energy = cudaq::observe(ansatz_cpp{}, h_cpp_example);
  std::cout << "Observed energy (C++): " << energy << std::endl;
  // [End SpinOp Creation C++ Execution]
  return 0; // Indicate success
}
